### ISSUE TITLE
RRuntimeError: Unable to find an inherited method for function ‘assay’ for signature ‘"dgCMatrix", "character"’

### ISSUE THREAD

**Issue by dburkhardt** [theislab/scib#141]
I'm trying to troubleshoot a bug associate with a package I'm working on with @LuckyMD (https://github.com/singlecellopenproblems/SingleCellOpenProblems). I tried using toy data and am getting this error

```python
import anndata
import numpy as np
import scIB.preprocessing

adata = anndata.AnnData(np.random.normal(100,1, size=(3000,1000)))
scIB.preprocessing.normalize(adata)
```

raises

```
---------------------------------------------------------------------------
RRuntimeError                             Traceback (most recent call last)
<ipython-input-1-68c695ef7288> in <module>
      4
      5 adata = anndata.AnnData(np.random.normal(100,1, size=(3000,1000)))
----> 6 scIB.preprocessing.normalize(adata)

~/.local/lib/python3.8/site-packages/scIB/preprocessing.py in normalize(adata, min_mean)
    168     ro.globalenv['data_mat'] = adata.X.T
    169     ro.globalenv['input_groups'] = adata_pp.obs['groups']
--> 170     size_factors = ro.r(f'computeSumFactors(data_mat, clusters = input_groups, min.mean = {min_mean})')
    171     del adata_pp
    172

/usr/lib/python3.8/site-packages/rpy2/robjects/__init__.py in __call__(self, string)
    414     def __call__(self, string):
    415         p = rinterface.parse(string)
--> 416         res = self.eval(p)
    417         return conversion.rpy2py(res)
    418

/usr/lib/python3.8/site-packages/rpy2/robjects/functions.py in __call__(self, *args, **kwargs)
    195                 v = kwargs.pop(k)
    196                 kwargs[r_k] = v
--> 197         return (super(SignatureTranslatedFunction, self)
    198                 .__call__(*args, **kwargs))
    199

/usr/lib/python3.8/site-packages/rpy2/robjects/functions.py in __call__(self, *args, **kwargs)
    123             else:
    124                 new_kwargs[k] = conversion.py2rpy(v)
--> 125         res = super(Function, self).__call__(*new_args, **new_kwargs)
    126         res = conversion.rpy2py(res)
    127         return res

/usr/lib/python3.8/site-packages/rpy2/rinterface_lib/conversion.py in _(*args, **kwargs)
     42 def _cdata_res_to_rinterface(function):
     43     def _(*args, **kwargs):
---> 44         cdata = function(*args, **kwargs)
     45         # TODO: test cdata is of the expected CType
     46         return _cdata_to_rinterface(cdata)

/usr/lib/python3.8/site-packages/rpy2/rinterface.py in __call__(self, *args, **kwargs)
    622                     error_occured))
    623             if error_occured[0]:
--> 624                 raise embedded.RRuntimeError(_rinterface._geterrmessage())
    625         return res
    626

RRuntimeError: Error in (function (classes, fdef, mtable)  :
  unable to find an inherited method for function ‘assay’ for signature ‘"dgCMatrix", "character"’
```

I'm have the following package versions installed

```
anndata 0.7.4
scanpy 1.4.5.post3
rpy2 3.3.5
anndata2ri 1.0.3
scanorama 1.6
memory_profiler 0.57.0
networkx 2.4
igraph 0.8.2
louvain 0.6.1
```

**Comment by dburkhardt** [comment]
As a note, this is reproducible with real data

```python
from openproblems.tasks.label_projection.datasets import  zebrafish_labels
import scIB.preprocessing

adata = zebrafish_labels(test=False)
scIB.preprocessing.normalize(adata)
```
raises
```
---------------------------------------------------------------------------
RRuntimeError                             Traceback (most recent call last)
<ipython-input-16-d43d74662bad> in <module>
      3
      4 adata = zebrafish_labels(test=False)
----> 5 log_scran_pooling(adata)

~/.local/lib/python3.8/site-packages/openproblems/tools/utils.py in normalize(adata, obsm, obs, var, *args, **kwargs)
     22                 adata.X = adata.layers[func.__name__]
     23             else:
---> 24                 func(adata, *args, **kwargs)
     25                 adata.layers[func.__name__] = adata.X
     26

~/.local/lib/python3.8/site-packages/openproblems/tools/normalize.py in log_scran_pooling(adata)
     12     scprep.run.install_bioconductor("scran")
     13     # Normalize via scran-pooling with own clustering at res=0.5
---> 14     scIB.preprocessing.normalize(adata)
     15
     16     # Make lightweight

~/.local/lib/python3.8/site-packages/scIB/preprocessing.py in normalize(adata, min_mean)
    168     ro.globalenv['data_mat'] = adata.X.T
    169     ro.globalenv['input_groups'] = adata_pp.obs['groups']
--> 170     size_factors = ro.r(f'computeSumFactors(data_mat, clusters = input_groups, min.mean = {min_mean})')
    171     del adata_pp
    172

/usr/lib/python3.8/site-packages/rpy2/robjects/__init__.py in __call__(self, string)
    414     def __call__(self, string):
    415         p = rinterface.parse(string)
--> 416         res = self.eval(p)
    417         return conversion.rpy2py(res)
    418

/usr/lib/python3.8/site-packages/rpy2/robjects/functions.py in __call__(self, *args, **kwargs)
    195                 v = kwargs.pop(k)
    196                 kwargs[r_k] = v
--> 197         return (super(SignatureTranslatedFunction, self)
    198                 .__call__(*args, **kwargs))
    199

/usr/lib/python3.8/site-packages/rpy2/robjects/functions.py in __call__(self, *args, **kwargs)
    123             else:
    124                 new_kwargs[k] = conversion.py2rpy(v)
--> 125         res = super(Function, self).__call__(*new_args, **new_kwargs)
    126         res = conversion.rpy2py(res)
    127         return res

/usr/lib/python3.8/site-packages/rpy2/rinterface_lib/conversion.py in _(*args, **kwargs)
     42 def _cdata_res_to_rinterface(function):
     43     def _(*args, **kwargs):
---> 44         cdata = function(*args, **kwargs)
     45         # TODO: test cdata is of the expected CType
     46         return _cdata_to_rinterface(cdata)

/usr/lib/python3.8/site-packages/rpy2/rinterface.py in __call__(self, *args, **kwargs)
    622                     error_occured))
    623             if error_occured[0]:
--> 624                 raise embedded.RRuntimeError(_rinterface._geterrmessage())
    625         return res
    626

RRuntimeError: Error in (function (classes, fdef, mtable)  :
  unable to find an inherited method for function ‘assay’ for signature ‘"dgRMatrix", "character"’
```

**Comment by scottgigante** [comment]
This is reproducible in bioc 3.11 but not in 3.10. 
```
scran::computeSumFactors(matrix(runif(0, 100, 100), ncol=10))
```

In 3.10:
```
> scran::computeSumFactors(matrix(runif(10, 100, 100), ncol=10))
 [1] 1 1 1 1 1 1 1 1 1 1
```

In 3.11:
```
> scran::computeSumFactors(matrix(runif(10, 100, 100), ncol=10))
Error in (function (classes, fdef, mtable)  :
  unable to find an inherited method for function ‘assay’ for signature ‘"matrix", "character"’
```

**Comment by scottgigante** [comment]
```
> BiocManager::version()
[1] ‘3.11’
> sessionInfo()
```

<details>

```
R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Arch Linux

Matrix products: default
BLAS/LAPACK: /usr/lib/libopenblas_haswellp-r0.3.10.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
 [1] SummarizedExperiment_1.18.2 statmod_1.4.34
 [3] scran_1.16.0                beeswarm_0.2.3
 [5] locfit_1.5-9.4              tidyselect_1.1.0
 [7] BiocSingular_1.4.0          purrr_0.3.4
 [9] lattice_0.20-41             colorspace_1.4-1
[11] vctrs_0.3.4                 generics_0.0.2
[13] viridisLite_0.3.0           stats4_4.0.2
[15] rlang_0.4.7                 pillar_1.4.6
[17] glue_1.4.2                  BiocParallel_1.22.0
[19] SingleCellExperiment_1.10.1 BiocGenerics_0.34.0
[21] dqrng_0.2.1                 matrixStats_0.56.0
[23] GenomeInfoDbData_1.2.3      lifecycle_0.2.0
[25] zlibbioc_1.34.0             munsell_0.5.0
[27] gtable_0.3.0                rsvd_1.0.3
[29] Biobase_2.48.0              IRanges_2.22.2
[31] GenomeInfoDb_1.24.2         vipor_0.4.5
[33] irlba_2.3.3                 scater_1.16.2
[35] parallel_4.0.2              BiocNeighbors_1.6.0
[37] Rcpp_1.0.5                  edgeR_3.30.3
[39] scales_1.1.1                limma_3.44.3
[41] DelayedArray_0.14.1         S4Vectors_0.26.1
[43] XVector_0.28.0              gridExtra_2.3
[45] ggplot2_3.3.2               dplyr_1.0.2
[47] GenomicRanges_1.40.0        grid_4.0.2
[49] tools_4.0.2                 bitops_1.0-6
[51] magrittr_1.5                RCurl_1.98-1.2
[53] tibble_3.0.3                crayon_1.3.4
[55] pkgconfig_2.0.3             ellipsis_0.3.1
[57] Matrix_1.2-18               DelayedMatrixStats_1.10.1
[59] ggbeeswarm_0.6.0            viridis_0.5.1
[61] R6_2.4.1                    igraph_1.2.5
[63] compiler_4.0.2
```

</details>

```
> BiocManager::version()
[1] ‘3.10’
> sessionInfo()
```

<details>

```
R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5                  vipor_0.4.5
 [3] pillar_1.4.6                compiler_3.6.3
 [5] GenomeInfoDb_1.22.1         XVector_0.26.0
 [7] viridis_0.5.1               bitops_1.0-6
 [9] BiocNeighbors_1.4.2         tools_3.6.3
[11] DelayedMatrixStats_1.8.0    zlibbioc_1.32.0
[13] statmod_1.4.34              SingleCellExperiment_1.8.0
[15] viridisLite_0.3.0           tibble_3.0.3
[17] lifecycle_0.2.0             gtable_0.3.0
[19] lattice_0.20-41             pkgconfig_2.0.3
[21] rlang_0.4.7                 Matrix_1.2-18
[23] igraph_1.2.5                DelayedArray_0.12.3
[25] parallel_3.6.3              beeswarm_0.2.3
[27] gridExtra_2.3               GenomeInfoDbData_1.2.2
[29] scater_1.14.6               vctrs_0.3.2
[31] S4Vectors_0.24.4            IRanges_2.20.2
[33] stats4_3.6.3                locfit_1.5-9.4
[35] grid_3.6.3                  glue_1.4.1
[37] Biobase_2.46.0              R6_2.4.1
[39] BiocParallel_1.20.1         ggbeeswarm_0.6.0
[41] limma_3.42.2                irlba_2.3.3
[43] ggplot2_3.3.2               BiocSingular_1.2.2
[45] edgeR_3.28.1                magrittr_1.5
[47] ellipsis_0.3.1              scales_1.1.1
[49] matrixStats_0.56.0          BiocGenerics_0.32.0
[51] GenomicRanges_1.38.0        SummarizedExperiment_1.16.1
[53] rsvd_1.0.3                  colorspace_1.4-1
[55] dqrng_0.2.1                 munsell_0.5.0
[57] RCurl_1.98-1.2              crayon_1.3.4
[59] scran_1.14.6
```

</details>

**Comment by scottgigante** [comment]
This works in both:
```
BiocGenerics::sizeFactors(scran::computeSumFactors(SingleCellExperiment::SingleCellExperiment(list(counts=matrix(runif(10, 100, 100), ncol=10)))))
```

**Comment by LuckyMD** [comment]
So in version 3.11 `scran` requires an `sce` object as input. A pity, but okay... will test whether this is consistent in 3.10 as well on my end and then merge. Thanks @scottgigante !

**Comment by scottgigante** [comment]
I'd make a sassy comment but you already know how I feel about tools not accepting standard data formats as input ;)

**Comment by LuckyMD** [comment]
Haha, i don't mind as long as it's consistent tbh... but yeah :/

**Comment by LuckyMD** [comment]
> ```
> BiocGenerics::sizeFactors(
>   scran::computeSumFactors(
>     SingleCellExperiment::SingleCellExperiment(list(
>       counts=matrix(runif(10, 100, 100), ncol=10)
>     ))
>   )
> )
> ```

Why the `BiocGenerics::sizeFactors()` wrapping?

**Comment by scottgigante** [comment]
Otherwise it returns the sce rather than the sizeFactors vector.

**Comment by LuckyMD** [comment]
So it turns out I'm running bioconductor version 3.9. And:
```
size_factors = ro.r(f'computeSumFactors(SingleCellExperiment::SingleCellExperiment(list(data_mat)), clusters = input_groups, min.mean = {min_mean})')
```
give me the error:
```
RRuntimeError: Error in assay(x, i = assay.type) : 
  'assay(<SingleCellExperiment>, i="character", ...)' invalid subscript 'i'
'counts' not in names(assays(<SingleCellExperiment>))
Calls: <Anonymous> ... .local -> .computeSumFactors -> ncol -> assay -> assay
```

I really love bioconductor version updates :/

**Comment by scottgigante** [comment]
Can you try `SingleCellExperiment::SingleCellExperiment(list(counts=data_mat))`?

**Comment by LuckyMD** [comment]
damn, I thought that's what I was doing the whole time. So it seems as though they introduced a new function `calculateSumFactors()` which uses the data matrix input. `computeSumFactors()` now only takes `SingleCellExperiment` objects where it did both before. So either, move to `sce` for everything, or write a bioconductor version check to see if it's >=3.11.

**Comment by LuckyMD** [comment]
Okay, works now. Thanks a lot for catching that typo! The error message wasn't really helping and I clearly wasn't looking at what I put in there :/

**Comment by LuckyMD** [comment]
I just saw you had a PR for this already... no idea what I was doing there...

**Comment by scottgigante** [comment]
Lol no worries. Thanks for acting on this so quickly!

**Comment by LuckyMD** [comment]
Just merged your PR #157 after reverting mine... credit where credit's due...

**Comment by scottgigante** [comment]
hah thanks :D
